pkgs/_analyzer_macros/lib/analyzer_macro_implementation.dart

```diff
@@ -0,0 +1,163 @@
+  static Future<MacroImplementation> start({
+    required Uri packageConfig,
+    required Map<String, String> macroImplByName,
+  }) async {
+    final result = MacroImplementation._(packageConfig, macroImplByName);
+    result._host = await MacroHost.serve(
+        macroImplByName: macroImplByName, service: result);
+    return result;
+  }
+
+  @override
+  injected.MacroPackageConfigs get packageConfigs => MacroPackageConfigs(this);
+
+  @override
+  injected.MacroRunner get runner => MacroRunner(this);
+
+  @override
+  Future<Response?> handle(MacroRequest request) async {
+    if (request.type != MacroRequestType.queryRequest) return null;
+    return Response.queryResponse(QueryResponse());
+  }
+}
+
+class MacroPackageConfigs implements injected.MacroPackageConfigs {
+  final MacroImplementation impl;
+
+  MacroPackageConfigs(this.impl);
+
+  @override
+  bool isMacro(Uri uri, String name) =>
+      impl._host.isMacro(impl.packageConfig, QualifiedName('$uri#$name'));
+}
+
+class MacroRunner implements injected.MacroRunner {
+  final MacroImplementation impl;
+
+  MacroRunner(this.impl);
+
+  @override
+  injected.RunningMacro run(Uri uri, String name) => RunningMacro.run(
+      impl,
+      QualifiedName('$uri#$name'),
+      // Look up from the macro name to its implementation.
+      QualifiedName(impl.macroImplByName['$uri#$name']!));
+}
+
+class RunningMacro implements injected.RunningMacro {
+  final MacroImplementation impl;
+  final QualifiedName name;
+  final QualifiedName implementation;
+  late final Future _started;
+
+  RunningMacro._(this.impl, this.name, this.implementation);
+
+  static RunningMacro run(MacroImplementation impl, QualifiedName name,
+      QualifiedName implementation) {
+    final result = RunningMacro._(impl, name, implementation);
+    // TODO(davidmorgan): this is currently what starts the macro running,
+    // make it explicit.
+    result._started =
+        impl._host.queryMacroPhases(impl.packageConfig, implementation);
+    return result;
+  }
+
+  @override
+  Future<MacroExecutionResult> executeDeclarationsPhase(MacroTarget target,
+      DeclarationPhaseIntrospector declarationsPhaseIntrospector) async {
+    await _started;
+    return MacroExecutionResult(
+        target, await impl._host.augment(name, AugmentRequest(phase: 2)));
+  }
+
+  @override
+  Future<MacroExecutionResult> executeDefinitionsPhase(MacroTarget target,
+      DefinitionPhaseIntrospector definitionPhaseIntrospector) async {
+    await _started;
+    return MacroExecutionResult(
+        target, await impl._host.augment(name, AugmentRequest(phase: 3)));
+  }
+
+  @override
+  Future<MacroExecutionResult> executeTypesPhase(
+      MacroTarget target, TypePhaseIntrospector typePhaseIntrospector) async {
+    await _started;
+    return MacroExecutionResult(
+        target, await impl._host.augment(name, AugmentRequest(phase: 1)));
+  }
+}
+
+/// Converts [AugmentResponse] to [injected.MacroExecutionResult].
```

Comment:
Should we have TODOs for the empty responses here?

Reply:
This class is basically one big TODO :) added at the top.